### PR TITLE
Feature/force single drainage point 190

### LINKED ISSUE #190: Force a single drainage point for `fill_pits`
By default `fill_pits` treats pixels on the edge of a raster or that abut nodata that don't otherwise border a lower pixel to be a drain. This can lead to cases where a predetermined watershed polygon can result in a DEM that drains little tiny watersheds around the edge of the polygon. The biggest error with this is a single pixel in the Nile watershed which splits the nile so it is only 1/3 as long as it should be:

image

This is not an issue that can be fixed by burning streams or any other method except to force a single outlet to a watershed. This this issue proposes two enhancements:
* Add an optional argument to "fill pits" which takes an x/y raster coordinate to indicate that is the only drain in the watershed
* helper function in `pygeoprocessing.routing` to find either the lowest dem pixel that drains off the nodata/the edge of the raster OR the lowest pixel in the raster for cases that may be a large depression.

### PR DESCRIPTION
**Pull request by richpsharp**
This feature fixes #190 by adding a parameter to `fill_pits` called `single_outlet_tuple`. If present this forces the fill to treat that pixel as the only valid drain. Combined with another feature that `max_pixel_fill_count` can be set to -1 to fill non-stop this will force every non-nodata pixel in the raster to drain to the pixel targed in `single_outlet_tuple`. To help identify a candidate I've also added another function `detect_lowest_drain_and_sink` which returns the lowest edge drain and internal sink pixels that could be used in tandem with `single_outlet_tuple`. Here's all the changes:
* added new `single_outlet_tuple` parameter to `fill_pits`
* created new helper function `detect_lowest_drain_and_sink`
* fixed a bug in `fill_pits` that would have filled up the nodata region of a raster in a corner case
* added functionality that if `max_pixel_fill_count` is -1 it will not terminate early and set that to the default parameter for `fill_pits`
* added test cases to cover the above
* updated history describing the above

### CODE REVIEW

File: src/pygeoprocessing/routing/routing.pyx [hunk 1]
```diff
@@ -696,6 +700,13 @@ def fill_pits(
     # have already been processed
     cdef int feature_id
 
+    # used to handle the case for single outlet mode
+
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx]: remove empty line to better relate the comment to the code?

File: src/pygeoprocessing/routing/routing.pyx [hunk 2]
```diff
@@ -793,10 +804,10 @@ def fill_pits(
...
 
         # search block for locally undrained pixels
-        for yi in range(1, win_ysize+1):
-            yi_root = yi-1+yoff
-            for xi in range(1, win_xsize+1):
-                xi_root = xi-1+xoff
+        for yi in range(win_ysize):
+            yi_root = yi+yoff
+            for xi in range(win_xsize):
+                xi_root = xi+xoff
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx:809]: I'm assuming this is where you "fixed a prior bug". Could you just comment why it was iterating from `1` before and why we don't want to do that now? I'd love to know the context for this fix / change.

File: src/pygeoprocessing/routing/routing.pyx [hunk 3]
```diff
@@ -850,8 +876,12 @@ def fill_pits(
                     xi_q = search_queue.front().xi
                     yi_q = search_queue.front().yi
                     search_queue.pop()
-                    if search_steps > max_pixel_fill_count:
+                    if (max_pixel_fill_count > 0 and
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx:878]: Okay, so if `max_pixel_fill_count == -1`, we don't want to come in here and quit the search early. We are riding it out until we find that single drain point. (just 'typing' out loud)

Comment (dcdenu4) [reply src/pygeoprocessing/routing/routing.pyx:878]: Well I guess if `single_outlet` is `None`, then we're just searching until we find an actual drain, be it nodata, an edge, or an outlet.

Comment (dcdenu4) [reply src/pygeoprocessing/routing/routing.pyx:878]: Or if it's determined it should be filled.

File: src/pygeoprocessing/routing/routing.pyx [hunk 4]
```diff
@@ -4159,6 +4216,131 @@ def calculate_subwatershed_boundary(
...
+
+    Args:
+        dem_raster_path_band (tuple): a raster/path band tuple to detect
+            sinks in.
+
+    Return:
+        (drain_pixel, drain_height, sink_pixel, sink_height) -
+            two (x, y) tuples with corresponding heights, first
+            list is for edge drains, the second is for pit sinks. The x/y
+            coordinate is in raster coordinate space and `height` is the
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx]: Should `height` be double ticks here?

File: src/pygeoprocessing/routing/routing.pyx [hunk 5]
```diff
@@ -611,7 +607,9 @@ def _generate_read_bounds(offset_dict, raster_x_size, raster_y_size):
 
 def fill_pits(
         dem_raster_path_band, target_filled_dem_raster_path,
-        working_dir=None, long max_pixel_fill_count=_MAX_PIXEL_FILL_COUNT,
+        working_dir=None,
+        long long max_pixel_fill_count=-1,
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx:611]: This is probably going to have an impact on InVEST models if the default was `500`. Is there an issue in the InVEST repo mentioning this change and the need to update the models?

File: src/pygeoprocessing/routing/routing.pyx [hunk 6]
```diff
@@ -4159,6 +4216,131 @@ def calculate_subwatershed_boundary(
...
+            LOGGER.info(
+                '(infer_sinks): '
+                f'{current_pixel} of {raster_x_size * raster_y_size} '
+                'pixels complete')
+
+        # search block for locally undrained pixels
+        for yi in range(1, win_ysize+1):
+            yi_root = yi-1+yoff
+            for xi in range(1, win_xsize+1):
+                xi_root = xi-1+xoff
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx]: Here we are doing the iteration with the 1 offset. Is this still intended given the change above in `fill_dem`?

File: src/pygeoprocessing/routing/routing.pyx [hunk 7]
```diff
@@ -4159,6 +4216,131 @@ def calculate_subwatershed_boundary(
...
+                # this value is set in case it turns out to be the root of a
+                # pit, we'll start the fill from this pixel in the last phase
+                # of the algorithm
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx]: Update this comment? I don't think we're doing any filling here.

File: src/pygeoprocessing/routing/routing.pyx [hunk 8]
```diff
@@ -4159,6 +4216,131 @@ def calculate_subwatershed_boundary(
...
+
+        if ctime(NULL) - last_log_time > _LOGGING_PERIOD:
+            last_log_time = ctime(NULL)
+            current_pixel = xoff + yoff * raster_x_size
+            LOGGER.info(
+                '(infer_sinks): '
+                f'{current_pixel} of {raster_x_size * raster_y_size} '
+                'pixels complete')
+
+        # search block for locally undrained pixels
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx]: and sinks

File: tests/test_routing.py [hunk 9]
```diff
@@ -1164,6 +1162,65 @@ def test_extract_straher_streams_watersheds_d8(self):
         watershed_vector = None
         watershed_layer = None
 
+    def test_single_drain_point(self):
+        """PGP.routing: test single_drain_point pitfill."""
+        dem_array = numpy.zeros((11, 11), dtype=numpy.float32)
+        dem_array[0, 0] = -1.0
+        dem_array[1:8, 1:8] = -2.0
+        dem_array[10, 7] = -4
```
Comment (dcdenu4) [inline tests/test_routing.py]: should this be `-4.0` for consistency?

File: tests/test_routing.py [hunk 10]
```diff
@@ -1164,6 +1162,65 @@ def test_extract_straher_streams_watersheds_d8(self):
...
+        dem_array[8:11, 8:11] = 2.0
+        dem_array[9:11, 9:11] = 1.0
+        dem_array[10, 10] = -7.0
+        dem_path = os.path.join(self.workspace_dir, 'dem.tif')
+        _array_to_raster(dem_array, None, dem_path)
+
+        # outlet tuple at 0,0, just drain one edge
+        expected_array_0_0 = numpy.copy(dem_array)
+        expected_array_0_0[1:8, 1:8] = -1
```
Comment (dcdenu4) [inline tests/test_routing.py:1179]: Just my own curiosity here. A plateau of all `-1` will drain to an outlet of also `-1`? I guess I'd expect these to be ever so slightly greater than `-1`.

Comment (dcdenu4) [reply tests/test_routing.py:1179]: By plateau I guess I mean pit in this instance.

File: tests/test_routing.py [hunk 11]
```diff
@@ -1164,6 +1162,65 @@ def test_extract_straher_streams_watersheds_d8(self):
...
+        expected_array_0_0[10, 7] = 0
+        expected_array_0_0[8:11, 8:11] = 2.0
+
+        # output tuple at 5,5, it's a massive pit so drain the edges
+        expected_array_5_5 = numpy.copy(dem_array)
+        expected_array_5_5[8:11, 8:11] = 2.0
+        expected_array_5_5[10, 7] = 0
```
Comment (dcdenu4) [inline tests/test_routing.py:1186]: We're just testing the algorithm works by draining everything to Sarlacc Pit?

File: tests/test_routing.py [hunk 12]
```diff
@@ -1164,6 +1162,65 @@ def test_extract_straher_streams_watersheds_d8(self):
...
+
+        for output_tuple, expected_array in [
+                ((0, 0), expected_array_0_0),
+                ((5, 5), expected_array_5_5),
+                ]:
```
Comment (dcdenu4) [inline tests/test_routing.py]: Could we have another test in here or separate that also manipulates the `max_fill_counter` variable using the same fake dem you've used here?

File: tests/test_routing.py [hunk 13]
```diff
@@ -1164,6 +1162,65 @@ def test_extract_straher_streams_watersheds_d8(self):
...
+            fill_path = os.path.join(self.workspace_dir, 'filled.tif')
+            pygeoprocessing.routing.fill_pits(
+                (dem_path, 1), fill_path,
+                single_outlet_tuple=output_tuple,
+                working_dir=self.workspace_dir)
+            result_array = pygeoprocessing.raster_to_numpy_array(fill_path)
+            numpy.testing.assert_almost_equal(result_array, expected_array)
+
+    def test_detect_lowest_drain_and_sink(self):
+        """PGP>routing: test detect_lowest_sink_and_drain."""
```
Comment (dcdenu4) [inline tests/test_routing.py]: >PGP.routing

**Review by dcdenu4 — CHANGES_REQUESTED** [review]
Hey @richpsharp I just gave this a look through. I followed it pretty well I think and didn't find anything glaring. I have a few comments though that should be minor.

File: src/pygeoprocessing/routing/routing.pyx (hunk 5 above)
Comment (richpsharp) [reply src/pygeoprocessing/routing/routing.pyx:611]: Sure, I made an issue here: https://github.com/natcap/invest/issues/560 won't affect any test data and it was a stopgap to only affect runs that were run on degenerate DEMs.

File: src/pygeoprocessing/routing/routing.pyx (hunk 1 above)
Comment (richpsharp) [reply src/pygeoprocessing/routing/routing.pyx]: removed in 5aaefd9

File: src/pygeoprocessing/routing/routing.pyx (hunk 2 above)
Comment (richpsharp) [reply src/pygeoprocessing/routing/routing.pyx:809]: I just noticed this loop ran from 1 to n+1 but then inside the parameter was always i-1. so just changed it from 0 to n then just use i. There's no functional change here.

File: src/pygeoprocessing/routing/routing.pyx (hunk 3 above)
Comment (richpsharp) [reply src/pygeoprocessing/routing/routing.pyx:878]: Correct.

File: src/pygeoprocessing/routing/routing.pyx (hunk 4 above)
Comment (richpsharp) [reply src/pygeoprocessing/routing/routing.pyx]: It wasn't referring to an actual parameter which is why i didn't have double ticks, just the suffix of those but I changed it in 0fd7e78 from `height` to _height so that's more obvious.

File: src/pygeoprocessing/routing/routing.pyx (hunk 8 above)
Comment (richpsharp) [reply src/pygeoprocessing/routing/routing.pyx]: changed in 02f5637

File: src/pygeoprocessing/routing/routing.pyx (hunk 6 above)
Comment (richpsharp) [reply src/pygeoprocessing/routing/routing.pyx]: changed in 40ad0cf

File: tests/test_routing.py (hunk 9 above)
Comment (richpsharp) [reply tests/test_routing.py]: changed in c05182d

File: tests/test_routing.py (hunk 10 above)
Comment (richpsharp) [reply tests/test_routing.py:1179]: Yes because there's a -1 pixel that pours off the edge, so to drain you only need the other pixels to be at the same height.

File: tests/test_routing.py (hunk 11 above)
Comment (richpsharp) [reply tests/test_routing.py:1186]: GOOD ONE

File: tests/test_routing.py (hunk 12 above)
Comment (richpsharp) [reply tests/test_routing.py]: I added two more cases that use a `max_pit_fill_count` of 1 that'll result in no pit filling at all. Added in 75a05f5

File: tests/test_routing.py (hunk 13 above)
Comment (richpsharp) [reply tests/test_routing.py]: fixed in ff64743e1148d989d559e87db6c767465d130753

File: src/pygeoprocessing/routing/routing.pyx (hunk 7 above)
Comment (richpsharp) [reply src/pygeoprocessing/routing/routing.pyx]: Removed in 1e3f6bf

**Comment by richpsharp** [comment]
> Hey @richpsharp I just gave this a look through. I followed it pretty well I think and didn't find anything glaring. I have a few comments though that should be minor.

Thanks @dcdenu4, I think I addressed all your comments. Can you take a second look when you get a chance?

**Review by dcdenu4 — APPROVED** [review]
Thanks Rich! Looks okay to me.
